stoix/systems/q_learning/ff_noisy_dqn.py

```diff
@@ -0,0 +1,569 @@
+        (env_states, timesteps, keys), traj_batch = jax.lax.scan(
+            _env_step, (env_states, timesteps, keys), None, config.system.warmup_steps
+        )
+
+        # Add the trajectory to the buffer.
+        buffer_states = buffer_add_fn(buffer_states, traj_batch)
+
+        return env_states, timesteps, keys, buffer_states
+
+    batched_warmup_step: Callable = jax.vmap(
+        warmup, in_axes=(0, 0, 0, 0), out_axes=(0, 0, 0, 0), axis_name="batch"
+    )
+
+    return batched_warmup_step
+
+
+def get_learner_fn(
+    env: Environment,
+    q_apply_fn: ActorApply,
+    q_update_fn: optax.TransformUpdateFn,
+    buffer_fns: Tuple[Callable, Callable],
+    config: DictConfig,
+) -> LearnerFn[OffPolicyLearnerState]:
+    """Get the learner function."""
+
+    buffer_add_fn, buffer_sample_fn = buffer_fns
+
+    def _update_step(
+        learner_state: OffPolicyLearnerState, _: Any
+    ) -> Tuple[OffPolicyLearnerState, Tuple]:
+        def _env_step(
+            learner_state: OffPolicyLearnerState, _: Any
+        ) -> Tuple[OffPolicyLearnerState, Transition]:
+            """Step the environment."""
+            q_params, opt_states, buffer_state, key, env_state, last_timestep = learner_state
+
+            # SELECT ACTION
+            key, policy_key = jax.random.split(key)
+            actor_policy = q_apply_fn(q_params.online, last_timestep.observation)
+            action = actor_policy.sample(seed=policy_key)
+
+            # STEP ENVIRONMENT
+            env_state, timestep = jax.vmap(env.step, in_axes=(0, 0))(env_state, action)
+
+            # LOG EPISODE METRICS
+            done = timestep.last().reshape(-1)
+            info = timestep.extras["episode_metrics"]
+            next_obs = timestep.extras["next_obs"]
+
+            transition = Transition(
+                last_timestep.observation, action, timestep.reward, done, next_obs, info
+            )
+
+            learner_state = OffPolicyLearnerState(
+                q_params, opt_states, buffer_state, key, env_state, timestep
+            )
+            return learner_state, transition
+
+        # STEP ENVIRONMENT FOR ROLLOUT LENGTH
+        learner_state, traj_batch = jax.lax.scan(
+            _env_step, learner_state, None, config.system.rollout_length
+        )
+
+        params, opt_states, buffer_state, key, env_state, last_timestep = learner_state
+
+        # Add the trajectory to the buffer.
+        buffer_state = buffer_add_fn(buffer_state, traj_batch)
+
+        def _update_epoch(update_state: Tuple, _: Any) -> Tuple:
+            """Update the network for a single epoch."""
+
+            def _q_loss_fn(
+                q_params: FrozenDict,
+                target_q_params: FrozenDict,
+                transitions: Transition,
+            ) -> jnp.ndarray:
+
+                q_tm1 = q_apply_fn(q_params, transitions.obs).preferences
+                q_t = q_apply_fn(target_q_params, transitions.next_obs).preferences
```

Comment:
I would then add rng_key as a argument to the q_loss_fn here which you use to split two noise keys for the q_apply_fns here.